### ISSUE TITLE
Im always getting response with "TRV not available"  when i send 2 commands for 2 devices.

### ISSUE THREAD

**Issue by Floyddotnet** [Floyddotnet/esp32_mqtt_eq3#2]
Also: Im always getting response with "TRV not available"  when i send 2 commands for 2 devices.

One device is working, one says not available

_Originally posted by @MeisterQ in https://github.com/Floyddotnet/esp32_mqtt_eq3/issues/1#issuecomment-468007518_

**Comment by Floyddotnet** [comment]
I get the same issue if the calor APP is still open on my phone. Can you check this?

**Comment by Floyddotnet** [comment]
I will add an additional tests to my testsuite at the weekend for this case and report then my results.
Hopefull i recieve my other ordered EQ3 TRV's until this. :-)

**Comment by MeisterQ** [comment]
> 
> 
> I get the same issue if the calor APP is still open on my phone. Can you check this?

App wasnt opend for weeks. Bluetooth on phone is disabled.

<german>
PS: Ich hab noch 2 weitere. 4 Stück sollten dann hoffentlich auch gehen :) Viel spass dabei. Bin gespannt. </german>

**Comment by softypit** [comment]
"TRV not available" is a response to an open failure of the BLE connection to the TRV. I get this fairly often with one of my TRVs which has poor signal. The ESP will look for beacons from the device and try to open the connection as soon as it sees one. The beacons occur every few seconds but the open timeout is quite short - I'm not sure if it is configurable. I think it only needs to miss one or two beacons before the open fails.

**Comment by Floyddotnet** [comment]
First update: I can confirm this issue. It's definitly a range problem.
If i place both TRVs side by side all work fine.
If i place the second TRV a little further away (e.g. 3 meters without direct line of sight) the second TRV didn't response in a half of my testcases. the first one (side by side with the ESP) does it, at the same time.

TL;DR: 2 TRVs side by side with the ESP works, Increasing the range produce this error sometimes.

I try to figure out how i can change the connection timeout.

**Comment by MeisterQ** [comment]
> 
> 
> First update: I can confirm this issue. It's definitly a range problem.
> If i place both TRVs side by side all work fine.
> If i place the second TRV a little further away (e.g. 3 meters without direct line of sight) the second TRV didn't response in a half of my testcases. the first one (side by side with the ESP) does it, at the same time.
> 
> TL;DR: 2 TRVs side by side with the ESP works, Increasing the range produce this error sometimes.
> 
> I try to figure out how i can change the connection timeout.

Hey,

my ESP32 is places about 10cm from my Fritzbox away. The TRVs are in a line of sight of about 20cm an 1.5m away from the ESP.

Did you changed any code and uploadet new files?

**Comment by Floyddotnet** [comment]
> Did you changed any code and uploadet new files?

Currently nothing importent (only debug outputs).

I this case, i get the following output:

```
I (1408217) EQ3_MAIN: Send next command
W (1408637) BT_APPL: bta_gattc_conn_cback() - cif=3 connected=0 conn_id=3 reason=0x003e
I (1408637) EQ3_MAIN: ESP_GATTC_DISCONNECT_EVT, reason = ESP_GATT_CONN_FAIL_ESTABLISH
E (1408637) EQ3_MAIN: open failed, status 133
```

**Comment by softypit** [comment]
I've just looked through the logs of my heating control application and I am also seeing these issues quite regularly. This isn't really a problem for me as the application checks for the status response matching the value set and if there is no correct response after a minute it resends the command.

Looking at the data I think all of my TRVs have occasions when they fail to connect. Some are within 2 metres of an ESP and some are more than 5 metres away. 

I wonder if interference is an issue rather than simply distance/signal strength. Perhaps a mechanism to re-queue the command within the ESP for a number of retries may be required.

**Comment by MeisterQ** [comment]
> 
> 
> I've just looked through the logs of my heating control application and I am also seeing these issues quite regularly. This isn't really a problem for me as the application checks for the status response matching the value set and if there is no correct response after a minute it resends the command.
> 
> Looking at the data I think all of my TRVs have occasions when they fail to connect. Some are within 2 metres of an ESP and some are more than 5 metres away.
> 
> I wonder if interference is an issue rather than simply distance/signal strength. Perhaps a mechanism to re-queue the command within the ESP for a number of retries may be required.

So, Floyddotnets .bins are not working for me. Im still using the last release of softypit atm.

How can we solve that problem?

**Comment by softypit** [comment]
How are you loading the .bins on to the ESP? I normally use 'make flash' which works. I can try your mechanism to see if I can figure out the problem.

**Comment by Floyddotnet** [comment]
> Perhaps a mechanism to re-queue the command within the ESP for a number of retries may be required.

That was my thought, too. I work currently on this.

**Comment by softypit** [comment]
> > Perhaps a mechanism to re-queue the command within the ESP for a number of retries may be required.
> 
> That was my thought, too. I work currently on this.

I spent some time yesterday looking at this. I have a very basic implementation in my working copy. It's not had much testing but I'll update my repo later so feel free to take a look.

**Comment by softypit** [comment]
> > > Perhaps a mechanism to re-queue the command within the ESP for a number of retries may be required.
> > 
> > 
> > That was my thought, too. I work currently on this.
> 
> I spent some time yesterday looking at this. I have a very basic implementation in my working copy. It's not had much testing but I'll update my repo later so feel free to take a look.

While doing some testing I noticed an issue with the command queue/sequencing. If a command is received via mqtt while a previous command is in the connection stage it is possible for the ble context to get reused to try a second connection. This causes the connection to fail and the first command in the list reports 'trv not available'.
This is all caused by the timer complete event being queued but the running flag being cleared when the alarm triggers. So the timer appeared to not be running and got restarted when in fact its previous alarm trigger had not yet been handled.
I've dropped some changes into my repo as a beta to fix this. It needs quite a bit of testing before it can be considered a release-candidate.
This also contains some retry code that will retry a command up to 3 times before reporting 'trv not available'. REQUEUE_RETRY controls whether the retry gets pushed to the end of the queue to allow other commands to get executed sooner or is simply retried before other commands. 
Ignore the version number - it's just a beta.

**Comment by Floyddotnet** [comment]
Locks very good .. thank you for this.

Even 8 meters distance and a thin wall between the ESP and the TRV work in the first attempt.

**Comment by MeisterQ** [comment]
> 
> 
> Locks very good .. thank you for this.
> 
> Even 8 meters distance and a thin wall between the ESP and the TRV work in the first attempt.

Can some of you create .bins and send me the files or upload them?

**Comment by Floyddotnet** [comment]
yes

**Comment by Floyddotnet** [comment]
https://github.com/Floyddotnet/esp32_mqtt_eq3/releases/tag/v1.32-dev

**Comment by MeisterQ** [comment]
> 
> 
> https://github.com/Floyddotnet/esp32_mqtt_eq3/releases/tag/v1.32-dev

Thank you very much. I will try them

**Comment by MeisterQ** [comment]
> 
> 
> https://github.com/Floyddotnet/esp32_mqtt_eq3/releases/tag/v1.32-dev

Is it me, or do i have to re-enter the data of my WiFi, mqttserver, etc.. after a re-powerup?

They are autofilled after a reset, but its not responding to my polls via mqtt.
When i then let the ESP go into AP mode, fill in all my data or just click on submit, its working fine.

**Comment by Floyddotnet** [comment]
strange. for me it worked without having to re-enter the data

**Comment by softypit** [comment]
> > https://github.com/Floyddotnet/esp32_mqtt_eq3/releases/tag/v1.32-dev
> 
> Is it me, or do i have to re-enter the data of my WiFi, mqttserver, etc.. after a re-powerup?
> 
> They are autofilled after a reset, but its not responding to my polls via mqtt.
> When i then let the ESP go into AP mode, fill in all my data or just click on submit, its working fine.

Does the ESP go into AP mode by itself or are you using the button to force it? If you have a console connection it should give more info.
I made some changes to the webserver in this beta so it always runs even when wifi is in STA mode. There are some unfinished status pages on there to show discovered devices and configuration. This change should not cause your issues.

**Comment by MeisterQ** [comment]
> 
> 
> > > https://github.com/Floyddotnet/esp32_mqtt_eq3/releases/tag/v1.32-dev
> > 
> > 
> > Is it me, or do i have to re-enter the data of my WiFi, mqttserver, etc.. after a re-powerup?
> > They are autofilled after a reset, but its not responding to my polls via mqtt.
> > When i then let the ESP go into AP mode, fill in all my data or just click on submit, its working fine.
> 
> Does the ESP go into AP mode by itself or are you using the button to force it? If you have a console connection it should give more info.
> I made some changes to the webserver in this beta so it always runs even when wifi is in STA mode. There are some unfinished status pages on there to show discovered devices and configuration. This change should not cause your issues.

Im forceing the ESP into AP mode, because i tried out if it needs the data to be manually filled in.

Well. I will take a look on this and will try it on a diffrent ESP.

**Comment by MeisterQ** [comment]
The latest beta release is working very fine. Since a week i hadnt any connection lost or no feedback.

**Comment by Floyddotnet** [comment]
Nice to hear. I also didn't have any stability problems. Great work @softypit
